### PR TITLE
JMS Connector Improvements

### PR DESCRIPTION
**Pull request by akalankapagoda**
The following improvements are introduced to the JMS connectors from this pull request.

1. Add Pooled JMS Connection Factory.
2. Add JMS Receiver for synchronous message receiving.
3. Improve error handling.

### CODE REVIEW

File: jms/org.wso2.carbon.transport.jms/pom.xml [hunk 1]
```diff
@@ -75,7 +79,7 @@
             org.slf4j.*;version="${slf4j.logging.package.import.version.range}",
             org.wso2.carbon.messaging.*;version="${carbon.messaging.package.import.version.range}",
             javax.jms.*;version="${javax.jms.import.version.range}",
-            org.wso2.carbon.connector.framework.*;version="${connector.framework.package.export.version}"
+            org.wso2.carbon.connector.framework.*;version="${connector.framework.package.export.version}",
```
Comment (isudana) [inline jms/org.wso2.carbon.transport.jms/pom.xml]: what's the reason for adding a ',' at the end ?

File: jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java [hunk 2]
```diff
@@ -129,14 +112,29 @@ public Connection getConnection(String userName, String password) throws JMSExce
...
-            cachedConnection = connection;
+            if (cachedConnection == null) {
+                connection = super.createConnection(userName, password);
+                cachedConnection = connection;
+            } else {
+                connection = cachedConnection;
+            }
+
+            try {
+                connection.start();
```
Comment (isudana) [inline jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java:126]: if it is a cached connection, do we need to start() ?

File: jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/PooledJMSConnectionFactory.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    @Override
+    public boolean validateObject(PooledConnectionKey key, PooledObject<Connection> pooledConnectionObject) {
+
+        boolean valid = false;
+
+        try {
+            pooledConnectionObject.getObject().start();
+            valid = true;
+        } catch (JMSException e) {
+            log.warn("Connection with username " + key + " is not valid anymore");
```
Comment (isudana) [inline jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/PooledJMSConnectionFactory.java]: Are we sure this is only happening for invalid username password ?

**Review by isudana — CHANGES_REQUESTED** [review]

File: jms/org.wso2.carbon.transport.jms/pom.xml (hunk 1 above)
Comment (akalankapagoda) [reply jms/org.wso2.carbon.transport.jms/pom.xml]: This is a mistake. Thanks for pointing out, will fix.

File: jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java (hunk 2 above)
Comment (akalankapagoda) [reply jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java:126]: A call to Connection.start on a connection that has already been started is ignored.

File: jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/PooledJMSConnectionFactory.java (hunk 3 above)
Comment (akalankapagoda) [reply jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/PooledJMSConnectionFactory.java]: We cannot be sure of this, hence will change the message. Thanks for pointing this out.

**Review by bsenduran — APPROVED** [review]

**Pull request merged**